src/components/MultiHopTrade/components/TradeConfirm/helpers.ts

```diff
@@ -0,0 +1,147 @@
+import { HopExecutionState } from 'state/slices/tradeQuoteSlice/types'
+
+export const getHopExecutionStateButtonTranslation = (hopExecutionState: HopExecutionState) => {
+  switch (hopExecutionState) {
+    case HopExecutionState.Pending:
+      return 'trade.pending'
+    case HopExecutionState.AwaitingAllowanceReset:
+      return 'common.reset'
+    case HopExecutionState.AwaitingAllowanceApproval:
+      return 'common.approve'
+    case HopExecutionState.AwaitingPermit2:
+      return 'trade.permit2.signMessage'
+    case HopExecutionState.AwaitingSwap:
+      return 'trade.signAndSwap'
+    case HopExecutionState.Complete:
+      return 'trade.doAnotherTrade'
+    default:
+      assertUnreachable(hopExecutionState)
+  }
+}
+
+export const getHopExecutionStateSummaryStepTranslation = (
+  hopExecutionState: HopExecutionState,
+  swapperName: string,
+): TextPropTypes['translation'] | null => {
+  switch (hopExecutionState) {
+    case HopExecutionState.Pending:
+      return null // No summary step for pending state
+    case HopExecutionState.AwaitingAllowanceReset:
+      return 'trade.awaitingAllowanceReset'
+    case HopExecutionState.AwaitingAllowanceApproval:
+      return 'trade.awaitingApproval'
+    case HopExecutionState.AwaitingPermit2:
+      return 'trade.awaitingPermit2Approval'
+    case HopExecutionState.AwaitingSwap:
+      return ['trade.awaitingSwap', { swapperName }]
+    case HopExecutionState.Complete:
+      return 'trade.complete'
+    default:
+      assertUnreachable(hopExecutionState)
+  }
+}
+
+type TradeStepParams = {
+  firstHopAllowanceApproval: ApprovalExecutionMetadata
+  firstHopPermit2: Omit<ApprovalExecutionMetadata, 'txHash' | 'isInitiallyRequired'> & {
+    permit2Signature?: string | undefined
+  }
+  firstHopAllowanceReset: ApprovalExecutionMetadata
+  lastHopAllowanceApproval: ApprovalExecutionMetadata
+  lastHopPermit2: Omit<ApprovalExecutionMetadata, 'txHash' | 'isInitiallyRequired'> & {
+    permit2Signature?: string | undefined
+  }
+  lastHopAllowanceReset: ApprovalExecutionMetadata
+  isMultiHopTrade?: boolean
+}
+
+export enum TradeStep {
+  FirstHopReset = 'firstHopReset',
+  FirstHopApproval = 'firstHopApproval',
+  FirstHopSwap = 'firstHopSwap',
+  LastHopReset = 'lastHopReset',
+  LastHopApproval = 'lastHopApproval',
+  LastHopSwap = 'lastHopSwap',
+  TradeComplete = 'tradeComplete',
+}
+
+export const getTradeSteps = (params: TradeStepParams): Record<TradeStep, boolean> => {
+  const {
+    firstHopAllowanceReset,
+    firstHopAllowanceApproval,
+    firstHopPermit2,
+    lastHopAllowanceReset,
+    lastHopAllowanceApproval,
+    lastHopPermit2,
+    isMultiHopTrade,
+  } = params
+  return {
+    [TradeStep.FirstHopReset]:
+      firstHopAllowanceReset.isRequired === true || firstHopAllowanceReset.txHash !== undefined,
+    [TradeStep.FirstHopApproval]:
+      firstHopAllowanceApproval.isInitiallyRequired === true ||
+      firstHopPermit2.isRequired === true ||
```

Comment:
q: is `isRequired` a catch all for permit2 (allowance grant for permit2 + actual permit2 message signing).

Feels odd that this all lives under the umbrella of `FirstHopApproval`, since users could have granted allowance to the permit2 contract but still require permit2 EIP-712 signature, in this case, there's no approval needed per se, but as far as the state machine is concerned, we still think there is.

Could be worth splitting the two in a follow-up, though obviously not a now thing, and not related to the trade flow improvement itself, this should work as-is as this is already the way in develop

cc @woodenfurniture